jmespath-core/src/main/java/io/burt/jmespath/jcf/JcfAdapter.java

```diff
@@ -143,6 +154,15 @@ public Object getProperty(Object value, String name) {
   }
 
   @Override
+  public Collection<String> getPropertyNames(Object value) {
+    if (isObject(value)) {
+      return ((Map<String, Object>) value).keySet();
```

Comment:
The JCF implementation is not meant to be used with structures that were created outside of this library. I think maybe that's where the confusion in this discussion lies. I don't expect anyone to build structures of `Map`, `List`, `String` and query them using JMESPath, that would be pretty insane. An adapter for more general Java object structures would have to work completely differently.